src/CHIPBackend.cc

```diff
@@ -1189,30 +1184,36 @@ void chipstar::Backend::trackEvent(
 }
 
 void chipstar::Backend::waitForThreadExit() {
-  /**
-   * If the main thread just creates a bunch of other threads and tries to exit
-   * right away, it could be the case that all those threads are not yet done
-   * with initialization. In particular, these threads might not have yet
-   * created their per-thread queues which is how we keep track of threads.
-   *
-   * So we just wait for 0.5 seconds before starting to check for thread exit.
-   */
+  // first, we must delay the main thread so that at least all other threads
+  // have gotten past
+  // libCHIP.so!chipstar::Device::getPerThreadDefaultQueueNoLock
+  // libCHIP.so!chipstar::Backend::findQueue
+  // libCHIP.so!hipMemcpyAsyncInternal
+  // libCHIP.so!hipMemcpyAsync
   pthread_yield();
-  // TODO fix-255 is there a better way to do this?
   unsigned long long int sleepMicroSeconds = 500000;
   usleep(sleepMicroSeconds);
 
-  while (true) {
-    {
-      auto NumPerThreadQueuesActive = ::Backend->getPerThreadQueuesActive();
-      if (!NumPerThreadQueuesActive)
+  // go through all devices checking their NumQueuesAlive until all they're all
+  // 1 or 0 (0 would indicate that hipStreamPerThread was never used and 1 would
+  // indicate main thread used hipStreamPerThread)
+  bool AllThreadsExited = false;
+  while (!AllThreadsExited) {
+    AllThreadsExited = true;
+    for (auto &Dev : getDevices()) {
+      if (Dev->NumQueuesAlive.load(std::memory_order_relaxed) > 1) {
+        AllThreadsExited = false;
         break;
+      }
+    }
 
-      logDebug("Backend::waitForThreadExit() per-thread queues still active "
-               "{}. Sleeping for 1s..",
-               NumPerThreadQueuesActive);
+    // logDebug and wait for 1 second
+    if (!AllThreadsExited) {
+      logWarn("Waiting for all per-thread queues to exit... This condition "
+              "would indicate that the main thread didn't call "
+              "join()");
+      sleep(1);
```

Comment:
The sleep here feels wrong/hacky. Could it be a condition variable invoked when the device's ref count goes to zero?

Reply:
That's what we do above - we have an atomic counter in the queue ctor/dtor.

The issue the main thread can exit before all the threads have reached the ctor. I agree that this is not optimal but I can't think of a better way to solve this. I tried creating another thread counter class and making it public thread_local hoping that as soon as a thread is created the constructor would get called but it still gets called later down in the callstack as indicated in the comment.

Perhaps there's a clang trick to force a certain object constructor to get execute before others? @pjaaskel

Reply:
Sorry I thought you meant the first sleep statement but my question still stands.

As for this, this sleep is part of the look where we check the counter. I think we want this debug statement and if we remove the sleep here it will spam stdout. This section of the code is only run in the case that a user wrote a bad hip program where threads are not joined() explicitly.

Reply:
Yes, I just meant to have a cleaner approach for the wait instead of sleep-polling with a rather long sleep time - a (pthread) condition variable kind of approach. Not critical if you think it's good enough.